### ISSUE TITLE
_mobius_matvec

### ISSUE THREAD

**Issue by GitZH-Chen** [geoopt/geoopt#233]
Bug: Incorrect use of `.prod(..., dtype=torch.bool)` in `_mobius_matvec` causes invalid fallback condition in high dimensions

Description:

In the current implementation of `_mobius_matvec`, the fallback condition is written as:

    cond = (mx == 0).prod(dim=dim, keepdim=True, dtype=torch.bool)

This line can incorrectly return `True` even when all entries of `mx` are non-zero, especially when `mx` has large dimensions (e.g., 128 or more). This leads to incorrect output by triggering the fallback case and selecting `res_0` instead of the correct Möbius result `res_c`.

How to Reproduce:

    import torch

    mx = torch.randn(2665, 128, dtype=torch.float64)
    print((mx == 0).sum())  # Expect 0
    cond = (mx == 0).prod(dim=1, keepdim=True, dtype=torch.bool)
    print(cond.unique())    # Expected: tensor([False]); Got: tensor([True]) ❌

Root Cause:

`.prod(..., dtype=torch.bool)` on boolean tensors is not a safe replacement for logical `.all()`. PyTorch may internally interpret this as a numeric product followed by a type cast, and in high dimensions this leads to incorrect `True` outputs.

Recommended Fix:

Replace:

    cond = (mx == 0).prod(dim=dim, keepdim=True, dtype=torch.bool)

with:

    cond = (mx == 0).all(dim=dim, keepdim=True)

This correctly returns `True` only when all elements along the dimension are exactly zero.

Suggested Fix Implementation:

    @torch.jit.script
    def _mobius_matvec(m: torch.Tensor, x: torch.Tensor, k: torch.Tensor, dim: int = -1):
        if m.dim() > 2 and dim != -1:
            raise RuntimeError("broadcasted Möbius matvec is supported for the last dim only")
        x_norm = x.norm(dim=dim, keepdim=True, p=2).clamp_min(1e-15)
        if dim != -1 or m.dim() == 2:
            mx = torch.tensordot(x, m, ([dim], [1]))
        else:
            mx = torch.matmul(m, x.unsqueeze(-1)).squeeze(-1)
        mx_norm = mx.norm(dim=dim, keepdim=True, p=2).clamp_min(1e-15)
        res_c = tan_k(mx_norm / x_norm * artan_k(x_norm, k), k) * (mx / mx_norm)
        cond = (mx == 0).all(dim=dim, keepdim=True)  # ✅ fixed condition
        res_0 = torch.zeros(1, dtype=res_c.dtype, device=res_c.device)
        res = torch.where(cond, res_0, res_c)
        return res

**Comment by SauravMaheshkar** [comment]
Makes sense, do you want to open a PR and contribute a fix 😃 ?
